thread/thread.h

```diff
@@ -362,17 +362,51 @@ namespace photon
         {
             return waitq::wait(timeout);
         }
+        template <typename LOCK, typename PRED,
+                  typename = decltype(std::declval<PRED>()())>
+        int wait(LOCK&& lock, PRED&& pred, Timeout timeout = {}) {
+            return do_wait_pred(
+                [&] { return wait(std::forward<LOCK>(lock), timeout); },
+                std::forward<PRED>(pred), timeout);
+        }
+        template <typename PRED,
+                  typename = decltype(std::declval<PRED>()())>
+        int wait_no_lock(PRED&& pred, Timeout timeout = {}) {
+            return do_wait_pred(
+                [&] { return wait_no_lock(timeout); },
+                std::forward<PRED>(pred), timeout);
+        }
         thread* signal()     { return resume_one(); }
         thread* notify_one() { return resume_one(); }
         int notify_all()     { return resume_all(); }
         int broadcast()      { return resume_all(); }
+    protected:
+        template<typename DO_WAIT, typename PRED>
+        int do_wait_pred(DO_WAIT&& do_wait, PRED&& pred, Timeout timeout) {
+            int ret = 0;
+            int err = ETIMEDOUT;
+            while (!pred() && !timeout.expired()) {
+                ret = do_wait();
+                err = errno;
+            }
+            errno = err;
+            return ret;
+        }
     };
 
     class semaphore : protected waitq
     {
     public:
         explicit semaphore(uint64_t count = 0) : m_count(count) { }
         int wait(uint64_t count, Timeout timeout = {});
```

Comment:
Is there any invocations of wait() that should be changed to invocations of wait_uninterruptible()?

Reply:
Make `wait` uninterruptible, keeps it works like it used to be, and add a new method `wait_interruptible` to present a wait that able to be interrupt might be better.